q2_metadata/normalization/_norm_check_rules.py

```diff
@@ -0,0 +1,311 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2016-2020, QIIME 2 development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file LICENSE, distributed with this software.
+# ----------------------------------------------------------------------------
+
+from q2_metadata.normalization._norm_errors import (
+    RuleError, ExpectedError, OntologyError, RemapError,
+    ValidationError, NormalizationError,
+    BlankError, MissingError, FormatError
+)
+
+
+def check_rule(variable: str, rules: dict, rule: str, rule_value):
+    """For each rule in the current variable's rules set,
+    perform checks that the rule values to be used for the rules
+    to apply are properly formatted.
+        - If properly formatted: full the rules dict
+        for the variable's Rules() class instance.
+        - If not properly formatted: raise a specific error and
+        collect the encountered error in the ErrorsCollection()
+        instance for this variable's Rules() class instance.
+
+    Parameters
+    ----------
+    variable : str
+        Name of the metadata variable for which there is rules.
+    rules : dict
+        Rules in the hard-coded data structure that has
+        common default values for all variables and
+        that is checked and updated here.
+    rule : str
+        Current rule name. Could be one of these:
+        - expected: accepted, controlled vocabulary from AGP.
+        - ontology: accepted, controlled vocabulary from a third-party resource.
+        - remap: the replacements to perform ( "A" -> "B" )
+        - validation: the
+        - normalization:
+        - blank:
+        - missing:
+        - format:
+    rule_value : str, list or dict
+        Parsed rule value for the current rule.
+    """
+    rule_type_map = {
+        'expected': 'lookups',
+        'ontology': 'lookups',
+        'remap': 'edits',
+        'validation': 'edits',
+        'normalization': 'edits',
+        'blank': 'allowed',
+        'missing': 'allowed',
+        'format': 'allowed'
+    }
+
+    if rule not in rule_type_map:
+        raise RuleError(rule, variable, rule_value)
+
+    if rule in ['blank', 'missing']:
+        rule_method_key = 'check_blank_missing'
+    else:
+        rule_method_key = 'check_%s' % rule
+
+    rule_method = globals()[rule_method_key]
+    rule_error = globals()['%sError' % rule.capitalize()]
+    rule_type = rule_type_map[rule]
+    if rule_method(rule_value):
+        rules[rule_type][rule] = rule_value
+    else:
+        raise rule_error(variable, rule_value)
+
+
+def check_expected(rule_value):
```

Comment:
It looks like this verifies that everything in a `list` is a `str`? I'm not seeing where this is used... is this function necessary in this PR?

Reply:
this is the function name that is "found" by the above `rule_method = globals()[rule_method_key]`.
This have been changed anyway as above review, and it is now more obvious in:
```
    rule_type_map = {
        'expected': (check_expected, ExpectedError, 'lookups', ),
        'ontology': (check_ontology, OntologyError, 'lookups'),
        'remap': (check_remap, RemapError, 'edits'),
        'validation': (check_validation, ValidationError, 'edits'),
        'normalization': (check_normalization, NormalizationError, 'edits'),
        'blank': (check_blank_missing, BlankError, 'allowed'),
        'missing': (check_blank_missing, MissingError, 'allowed'),
        'format': (check_format, FormatError, 'allowed')
    }

    if rule not in rule_type_map:
        raise RuleError(rule, variable, rule_value)

    rule_method, rule_error, rule_type = rule_type_map[rule]
    if rule_method(rule_value):
        rules[rule_type][rule] = rule_value
    else:
        raise rule_error(variable, rule_value)

```